File: java/src/main/java/com/sasteval/vuln/hard/DeserializationHandler.java
package com.sasteval.vuln.hard;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.ObjectInputStream;

/**
 * CWE-502: Deserialization of Untrusted Data
 * VULNERABILITY: Deserializes arbitrary bytes with no type filter or allowlist.
 */
public class DeserializationHandler {

    public static Object deserialize(byte[] data) throws IOException, ClassNotFoundException {
        // VULN: No ObjectInputFilter, no type checking — arbitrary class instantiation
        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(data));
        return ois.readObject();
    }
}

File: dotnet/Vuln/Hard/DeserializationHandler.cs
#pragma warning disable SYSLIB0011 // BinaryFormatter is obsolete

using System.Runtime.Serialization.Formatters.Binary;

namespace SastEval.Vuln.Hard;

/// <summary>
/// CWE-502: Deserialization of Untrusted Data - BinaryFormatter with no type binder.
/// </summary>
public static class DeserializationHandler
{
    public static object? Deserialize(byte[] data)
    {
        if (data == null || data.Length == 0)
        {
            return null;
        }

        // VULNERABLE: BinaryFormatter with NO type binder allows arbitrary type instantiation
        var formatter = new BinaryFormatter();
        using var stream = new MemoryStream(data);
        return formatter.Deserialize(stream);
    }
}

#pragma warning restore SYSLIB0011
